src/spc/sema.jl

```diff
@@ -0,0 +1,631 @@
+    if devkind == :d
+        return GlobalRef(CedarSim.SpectreEnvironment, :diode)
+    elseif devkind == :r
+        return GlobalRef(CedarSim.SpectreEnvironment, :resistor)
+    elseif devkind == :c
+        return GlobalRef(CedarSim.SpectreEnvironment, :capacitor)
+    end
+    if dialect == :ngspice
+        if devkind in (:pmos, :nmos)
+            if level == 5
+                file = stmt.ps.srcfile.path
+                line = SpectreNetlistParser.LineNumbers.compute_line(stmt.ps.srcfile.lineinfo, stmt.startof)
+                @warn "BSIM2 ($devkind at level $level) not implemented" _file=file _line=line
+                return GlobalRef(CedarSim, :UnimplementedDevice)
+            elseif level == 8 || level == 49
+                file = stmt.ps.srcfile.path
+                line = SpectreNetlistParser.LineNumbers.compute_line(stmt.ps.srcfile.lineinfo, stmt.startof)
+                @warn "BSIM3 ($devkind at level $level) not implemented" _file=file _line=line
+                return GlobalRef(CedarSim, :UnimplementedDevice)
+            elseif level == 14 || level == 54
+                return GlobalRef(sema_resolve_import(sema, :BSIM4), :bsim4)
+            elseif level == 17 || level == 72
+                if version == 107 || version === nothing
+                    return :bsimcmg107
+                else
+                    file = stmt.ps.srcfile.path
+                    line = SpectreNetlistParser.LineNumbers.compute_line(stmt.ps.srcfile.lineinfo, stmt.startof)
+                    @warn "Version $version of mosfet $devkind at level $level not implemented" _file=file _line=line
+                    return GlobalRef(CedarSim, :UnimplementedDevice)
+                end
+            else
+                file = stmt.ps.srcfile.path
+                line = SpectreNetlistParser.LineNumbers.compute_line(stmt.ps.srcfile.lineinfo, stmt.startof)
+                @warn "Mosfet $devkind at level $level not implemented" _file=file _line=line
+                return GlobalRef(CedarSim, :UnimplementedDevice)
+            end
+        elseif devkind in (:pnp, :npn)
+            file = stmt.ps.srcfile.path
+            line = SpectreNetlistParser.LineNumbers.compute_line(stmt.ps.srcfile.lineinfo, stmt.startof)
+            @warn "Bipolar $devkind at level $level not implemented" _file=file _line=line
+            return GlobalRef(CedarSim, :UnimplementedDevice)
+        elseif devkind == :sw
+            return GlobalRef(CedarSim.SpectreEnvironment, :Switch)
+        end
+    end
+    # Search for this model in the HDL Imports
+    if sema.parse_cache !== nothing
+        # Might be an import instead, which we will resolve later
+        return GlobalRef(sema.parse_cache.thismod, devkind)
+    end
+    file = stmt.ps.srcfile.path
+    line = SpectreNetlistParser.LineNumbers.compute_line(stmt.ps.srcfile.lineinfo, stmt.startof)
+    @warn "Device $devkind at level $level not implemented" _file=file _line=line
```

Comment:
Huzzah! Finally some sane model resolution errors.

Do we have an equivalent for parameters?

Reply:
No, but sema is a natural place to check these things.